src/derived.rs

```diff
@@ -228,6 +233,59 @@ impl<DB: Database> std::fmt::Debug for MemoInputs<DB> {
     }
 }
 
+type LinkedHashSet<T> = LinkedHashMap<T, (), BuildHasherDefault<FxHasher>>;
+
+struct QueryMap<DB, Q>
+where
+    Q: QueryFunction<DB>,
+    DB: Database,
+{
+    lru_keys: LinkedHashSet<Q::Key>,
+    data: FxHashMap<Q::Key, QueryState<DB, Q>>,
+}
+
+impl<DB, Q> Default for QueryMap<DB, Q>
+where
+    Q: QueryFunction<DB>,
+    DB: Database,
+{
+    fn default() -> Self {
+        QueryMap {
+            lru_keys: Default::default(),
+            data: Default::default(),
+        }
+    }
+}
+
+impl<DB, Q> QueryMap<DB, Q>
+where
+    Q: QueryFunction<DB>,
+    DB: Database,
+{
+    fn set_lru_capacity(&mut self, new_capacity: usize) {
```

Comment:
what happens if this gets invoked with `set_lru_capacity(0)`? seems like it will remove all of the keys, which probably isn't what we want?

Reply:
Ah, good catch!

Reply:
So, let's talk through why we use `usize` here and not an `Option<usize>`. I debated where I should go with `Option`, but decided to use special-cased zero instead. It seems like using 0 to mean "unlimited" is a common pattern in various configurations, and, for example, in rust-analyzer we can use `RA_LRU_CAPACITY=0` to disable LRU, instead of inventing some ad-hoc syntax like `RA_LRU_CAPACITY=none`. Note that this *could* be done on the UI level in rust-analyzer, by translating `0` into `None`, but I figured that just using `0` as a sentinel value internally will also work

Reply:
Yeah, it seemed fine to me. There's no useful interpretation to `0`.

Reply:
Well, I guess you could interpret it as "storage mode = deps only". But still I think using `0` as a sentinel value is ok.